glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt

```diff
@@ -173,9 +85,33 @@ class PingUploadWorker(context: Context, params: WorkerParameters) : Worker(cont
          * @return true if process was successful
          */
         internal fun uploadPings(): Boolean {
-            val storageDirectory = File(Glean.getDataDir(), PINGS_DIR)
+            var ffiTask = LibGleanFFI.INSTANCE.glean_get_upload_task()
+            var task = ffiTask.toPingUploadTask()
+            while (task is PingUploadTask.Upload) {
+                // Get the request
+                val request = task.request()!!
+                // Upload the request
+                // If the status is `null` there was some kind of unrecoverable error
+                // so we return 400 which will ensure this gets treated as such.
+                @Suppress("MagicNumber")
+                val status = Glean.httpClient.doUpload(
+                        request.path,
+                        request.body,
+                        request.headers,
+                        Glean.configuration
+                ) ?: 400
```

Comment:
We do, what I didn't account for when writing that `process_ping_upload_request` function is that sometimes we don't have a status code because we don't even attempt upload (e.g. malformed url error, i/o error). When we don't attempt upload, status code is null.

It is a hack and I am not super happy with the solution to turn `null` into `400`, so that the rust part treats it as an unrecoverable error, but it is the least disruptive to what we have now. Let me know if you have a better solution :) It is imperative that we call the process function even in these cases so that rust can delete that ping file.

Reply:
> 
> 
> We do, what I didn't account for when writing that `process_ping_upload_request` function is that sometimes we don't have a status code because we don't even attempt upload (e.g. malformed url error, i/o error). When we don't attempt upload, status code is null.

We can still change the Rust internals to match this behaviour. Maybe that's a blasphemy, but what about changing `status` [here](https://github.com/mozilla/glean/blob/b71434f4cd30729daf6433e25702a886971e7849/glean-core/src/upload/mod.rs#L178) to be a `Result<u16, Error>` and just pass the status code if `Ok` and let Rust properly handle the errors otherwise?

Reply:
This is a great suggestion :) I think it would work nicely, @badboy what do you think?

Reply:
This would require some decision making  as to how we would pass the `Error` through the FFI to Rust. `Result` is an enum and that is not the simplest to pass around. Overall I am not sure this would be worth the effort. One other option maybe would be to have status be an Option<u16>, that might be simpler to send through the FFI, but I would need to verify.

I will file a bug to investigate this further. Although the current solution to turn null into a known unrecoverable error is a bit hacky, it holds up until we have a proper non hacky solution, IMO.

Reply:
Done: https://bugzilla.mozilla.org/show_bug.cgi?id=1620682

Reply:
Please  prioritize taking that bug right after this.